src/main/java/io/github/oliviercailloux/y2018/apartments/valuefunction/profile/Profile.java

```diff
@@ -140,21 +123,15 @@ public LinearAVF getLinearAVF() {
   }
 
   /**
-   * Sets the linearAVF of a Profile, it needs to be the last argument setted to be able to check if
-   * the weight is in the range of all of the criterions.
+   * Sets the linearAVF of a <code>Profile</code> It's a clone of the current <code>Profile</code>,
+   * changing the <code>Profile</code> to <code> newLinearAvf </code>.
    *
+   * @param newLinearAvf recreate a Profile based on <code>newLinearAvf</code>.
    * @return Profile with its LinearAVF set
    */
   public Profile withLinearAVF(LinearAVF newLinearAvf) {
-    Profile prof = cloneProfile();
-
-    for (Criterion c : Criterion.values()) {
-      checkWeightInRange(c, newLinearAvf);
-    }
-
-    prof.linearAvf = newLinearAvf;
-
-    return prof;
+    Arrays.stream(Criterion.values()).forEach(c -> this.checkWeightInRange(c, newLinearAvf));
+    return clone(Optional.empty(), Optional.of(newLinearAvf));
```

Comment:
Au lieu d’utiliser 'clone`, il serait plus direct d’utiliser le nouveau constructeur. Passez-lui la map de critères existante et la nouvelle linearAvf.

La même approche devrait pouvoir s’appliquer partout où vous appelez clone.

Vous pouvez ensuite effacer clone.

FUTURE : se passer du Builder au niveau des utilisateurs de cette classe en utilisant plutôt la factory (à créer, voir ci-dessus) de paramètres avf et map. On devrait pouvoir ensuite effacer le builder, et simplifier largement cette classe, par exemple en mettant la map en immuable.

Reply:
Fix in 2716654f426de50b45f24b104fbd7e9972119556
Voir [commentaire ci-dessus](https://github.com/av1m/Apartments/pull/58#issuecomment-648685461)